Sage/carbon
# chore: update dependencies

### Proposed behaviour

#### Package removals
- Remove unused dependencies: `ts-node` and `esbuild`
- Remove dependency `@types/cypress` - since cypress provides its own in-built TypeScript types

#### Package upgrades

##### Major 
- Upgrade `@types/draft-js` to latest version
- Upgrade `@wojtekmaj/enzyme-adapter-react-17` to latest version
- Upgrade `husky` to the latest version
- Upgrade `commitlint` to the latest version
- Upgrade `@types/jest` to version 27.5
- Upgrade `@types/node` to version 18.16
- Upgrade `lint-staged` to version 11.2.6

##### Minor
- Upgrade `typescript` to version 4.7
- Limit `prettier` to version 2.2
- All other upgraded packages are now on their next latest minor version

##### Patch
- Restrict `@types/enzyme` to version 3.10.8

#### Other changes
- Update `npm run babel` script to reflect recent cypress changes from #5907

### Current behaviour

We have many dependencies that require upgrading.



### Checklist




### Additional context

Some package versions were restricted:

| package and restriction     | reasoning |
| ------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `lint-staged` restricted to version 11.x    | The package is a ESM-only module for versions >12 ([see release notes](https://github.com/okonet/lint-staged/releases/tag/v12.0.0)). We need to check if we can support this before upgrading in the future.                                                                                                                                          |
| `typescript`  restricted to version 4.7.x   | Before upgrading above version 4.7, the following must be done: 1. Modify types of `Button` props - to account for component's polymorphic nature, 2. Upgrade to storybook version 7 - to update an internal plugin that we depend on, which refers to deprecated TypeScript code ([see issue](https://github.com/storybookjs/storybook/issues/19618)) |
| `prettier`   restricted to version 2.2.x   | New formatting rules had been introduced in later minor versions, so this has been left to limit the number of changes in this PR                                                                                                                                                                                                                     |
| `@types/enzyme` restricted to version  3.10.8 | Later patches caused some type clashes between our React 17 types and the package's internal React 16 types. Therefore, package has been left on its current version, but it is worth highlighting this type clash as a problem if we plan to temporarily use an unofficial React 18 adapter.                                                                                                               |

### Testing instructions

- Check git hooks are still being triggered correctly on committing changes
   -  `commitlint` should be run before changes are committed
   -   `prettier` and `eslint` should be run before changes to JavaScript/TypeScript files are committed
- Check for no intentional changes to output from `npm run babel`
- Check there are no new chromatic changes
- Check storybook still builds correctly and runs locally
- Check all GitHub checks pass
